### PR TITLE
Replace the tvhxxx() macros with a separate logger utility

### PR DESCRIPTION
**Pull request by Jalle19**
@ksooo here it is. I'm not particularly proud of using a private singleton for storing the implementation method in a static context but for some reason I couldn't make `m_implementation` static. If you want to feel free to experiment with it.

Only managed to get rid of one `Tvheadend.h` include though, `Settings` still requires the `XBMC` pointer so we can't remove it yet (it could include `client.h` instead though).

### CODE REVIEW

File: src/tvheadend/utilities/Logger.h [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      ERROR,
+      INFO,
+      DEBUG,
+      TRACE
+    };
+
+    /**
+     * Short-hand for a function that acts as the logger implementation
+     */
+    typedef std::function<void(LogLevel level, const char *message)> LogImplementation;
```
Comment (Jalle19) [inline src/tvheadend/utilities/Logger.h]: Should probably be renamed to `LoggerImplementation`.

File: src/tvheadend/utilities/Logger.h [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      /**
+       * Logs the specified message using the specified log level
+       * @param level the log level
+       * @param message the log message
+       * @param ... parameters for the log message
+       */
+      static void Log(LogLevel level, const std::string &message, ...)
+      {
+        char buffer[MESSAGE_BUFFER_SIZE];
+        std::string logMessage = "pvr.hts - " + message;
```
Comment (Jalle19) [inline src/tvheadend/utilities/Logger.h]: In order to use this class in other PVR addons (I'm thinking about `pvr.vbox `mostly) this prefix should probably be configurable.

**Comment by ksooo** [comment]
Nice.

Regardig your comments to yourself - yep, makes sense ;-)

> ...but for some reason I couldn't make m_implementation static. If you want to feel free to experiment with it.

No time for experiments here atm. :-/

**Comment by Jalle19** [comment]
@ksooo I changed some semantics and made the log prefix configurable, plus rebased. In my opinion this is ready to go in. The risk for breakage is very low since there's no logic changes.

File: src/HTSPDemuxer.cpp [hunk 3]
```diff
@@ -417,7 +419,7 @@ void CHTSPDemuxer::ParseSubscriptionStart ( htsmsg_t *m )
            some versions of tvheadend and is here for backward compatibility. */
         if (stream.iWidth == 0 || stream.iHeight == 0)
         {
-          tvhinfo("Ignoring subscriptionStart, stream details missing");
+          Logger::Log(LogLevel::INFO, "Ignoring subscriptionStart, stream details missing");
```
Comment (ksooo) [inline src/HTSPDemuxer.cpp]: Not related to this PR, but why is this INFO, not DEBUG?

File: src/tvheadend/Subscription.cpp [hunk 4]
```diff
@@ -247,7 +249,7 @@ void Subscription::ParseSubscriptionStatus ( htsmsg_t *m )
   const char *status = htsmsg_get_str(m, "status");
 
   if (status != NULL)
-    tvhinfo("Bad subscription status: %s", status);
+    Logger::Log(LogLevel::INFO, "Bad subscription status: %s", status);
```
Comment (ksooo) [inline src/tvheadend/Subscription.cpp]: Same here. Why INFO, not DEBUG or ERROR?

**Comment by ksooo** [comment]
Like always, I'd love a changelog entry. ;-)

**Comment by Jalle19** [comment]
@ksooo added changelog entry. I can change the log level of those lines you mentioned in a separate PR once this is merged. No idea why the levels were set like this to begin with, the first one I suspect is a leftover from when I wanted to test it without having debugging enabled.

**Comment by ksooo** [comment]
> I can change the log level of those lines you mentioned in a separate PR once this is merged.

 That would be great.

**Comment by ksooo** [comment]
Thanks for another round of cleaning up the code.

**Pull request merged**